ssl/quic/quic_impl.c

```diff
@@ -1099,33 +1150,65 @@ static int quic_shutdown_wait(void *arg)
     return ossl_quic_channel_is_terminated(qc->ch);
 }
 
+/* Returns 1 if shutdown flush process has finished or is inapplicable. */
+static int quic_shutdown_flush_wait(void *arg)
+{
+    QUIC_CONNECTION *qc = arg;
+
+    return ossl_quic_channel_is_term_any(qc->ch)
+        || qc_shutdown_flush_finished(qc);
+}
+
 QUIC_TAKES_LOCK
 int ossl_quic_conn_shutdown(SSL *s, uint64_t flags,
                             const SSL_SHUTDOWN_EX_ARGS *args,
                             size_t args_len)
 {
     int ret;
     QCTX ctx;
+    int stream_flush = ((flags & SSL_SHUTDOWN_FLAG_NO_STREAM_FLUSH) == 0);
 
     if (!expect_quic(s, &ctx))
-        return 0;
+        return -1;
 
     if (ctx.is_stream)
         /* TODO(QUIC): Semantics currently undefined for QSSOs */
         return -1;
 
     quic_lock(ctx.qc);
 
+    if (ossl_quic_channel_is_terminated(ctx.qc->ch)) {
+        quic_unlock(ctx.qc);
+        return 1;
+    }
+
+    /* Phase 1: Stream Flushing */
+    if (stream_flush) {
+        qc_shutdown_flush_init(ctx.qc);
+
+        if (!qc_shutdown_flush_finished(ctx.qc)) {
+            if (qc_blocking_mode(ctx.qc))
+                block_until_pred(ctx.qc, quic_shutdown_flush_wait, ctx.qc, 0);
+            else
+                ossl_quic_reactor_tick(ossl_quic_channel_get_reactor(ctx.qc->ch), 0);
+        }
+
+        if (!qc_shutdown_flush_finished(ctx.qc)) {
+            quic_unlock(ctx.qc);
+            return 0; /* ongoing */
+        }
+    }
+
+    /* Phase 2: Connection Closure */
```

Comment:
IIUC at this point all our send streams have been flushed and we can no longer write any more data.  If there was no data to flush then we immediately proceed to connection closure without ever returning 0 in between. I'm not sure this is correct - or at least its not obviously correct. In TLS you can call SSL_shutdown(), receive a 0 return (as long as the peer didn't already initiate shutdown) and then continue to call SSL_read() until the peer sends close_notify. This doesn't look like it will work here because we proceed to immediate closure.

Reply:
I think its the continued ability to allow SSL_read() that I'm mostly concerned about (e.g. to receive already buffered data) then that is probably close enough??